[aotinductor] enable CUTLASS backend

ColinPeppler opened the pull request:
### Context
This PR allows CUTLASS kernels usage in AOTI. It does this by:
* For any CUTLASS kernels that win during autotuning, compile them as a .so & .o
* When creating the final model .so, link all the CUTLASS kernels .o files
* Make sure we codegen things correctly (argument dtypes and specify extern "C" linking for the CUTLASS kernel)

### Example
https://gist.github.com/ColinPeppler/e834fa2255c37e9444b6d540bf7bd04d#file-model-cpp-L548-L549

```
TORCH_LOGS="+output_code" python test/inductor/test_cutlass_backend.py -v -k test_max_autotune_cutlass_backend_regular_mm
```

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #134379



cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @amjames @desertfire @chauhang

ColinPeppler committed:
[wip] enable CUTLASS in AOTI

[ghstack-poisoned]

ColinPeppler committed:
Update on "[wip] enable CUTLASS in AOTI"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 amjames desertfire chauhang

[ghstack-poisoned]

ColinPeppler committed:
Update on "[wip] enable CUTLASS in AOTI"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 amjames desertfire chauhang

[ghstack-poisoned]

ColinPeppler committed:
Update on "[wip] enable CUTLASS in AOTI"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 amjames desertfire chauhang

[ghstack-poisoned]

ColinPeppler commented on torch/_inductor/codegen/cuda/cuda_cpp_scheduling.py:
```diff
@@ -66,7 +66,9 @@ def define_kernel(self, src_code: str, node_schedule) -> str:
             compile_wrapper.writeline("async_compile.cuda(r'''")
             compile_wrapper.splice(src_code, strip=True)
-            compile_wrapper.writeline("''', 'so')")
+            compile_wrapper.writeline(
+                f"''', 'so', aot_compile={str(V.graph.aot_mode)})"
+            )
```
When we compile kernels in the JIT step, then we also pass a flag to tell our CUDACodeCache to compile an object file for us. So, we can eventually link it.

I have a feeling this isn't the best way to do it, but this approach probably involves the least code changes.

ColinPeppler commented on torch/_inductor/kernel/mm.py:
```diff
@@ -180,7 +180,9 @@ def tuned_mm(mat1, mat2, *, layout=None):
             )
     if static_shape and is_nonzero and use_cutlass_template(layout, m, n, k):
-        CUTLASS3xGemmTemplate.add_cutlass_gemm_choices(choices, layout, [mat1, mat2])
+        # Developing on A100 rn.
+        CUTLASS2xGemmTemplate.add_cutlass_gemm_choices(choices, layout, [mat1, mat2])
+        # CUTLASS3xGemmTemplate.add_cutlass_gemm_choices(choices, layout, [mat1, mat2])
```
Need to figure out how to know when to use v2 vs v3.

chenyang78 replied:
Perhaps we could stick with the latest version, e.g. v3?

kadeng replied:
In general don't use v2. There's no performance benefit over Triton. I originally entirely removed Cutlass v2 support, but it was re-added since an external contribution required it for certain mixed-mode GEMMs that were using 2.x APIs.

ColinPeppler replied:
Thanks for the context Kai!

ColinPeppler commented on torch/_inductor/codecache.py:
```diff
@@ -1832,6 +1832,11 @@ def _pad_to_alignment(raw_bytes: bytes) -> bytes:
                 "linux": _compile_consts_linux,
                 "darwin": _compile_consts_darwin,
             }[sys.platform](aot_constants)
+            other_obj_files = []
+            for entry in CUDACodeCache.cache.values():
+                if entry.output_path.endswith(".o"):
+                    other_obj_files.append(entry.output_path)
+            other_o = " ".join(other_obj_files)
```
also pretty hacky imo.. maybe I can create another abstraction to easily grab object files for CUTLASS/CK wrapper functions.

tenpercent replied:
Is there a cache instance per model, or will it also pick up object files from different models?

ColinPeppler replied:
it's cached per model

and sg will add comments or think of a better name

tenpercent replied:
One caveat here is, the CK backend uses ROCmCodeCache, so there has to be a way to switch to a correct cache object

ColinPeppler commented on test/inductor/test_aot_inductor.py:
```diff
@@ -2650,6 +2650,33 @@ def grid(meta):
         )
+    
+
+    def test_cutlass_backend(self):
```
this is redundant lemme remove

tenpercent commented on test/inductor/test_cutlass_backend.py:
```diff
@@ -24,6 +24,12 @@
 
 
+try:
+    from .test_aot_inductor import AOTIRunnerUtil
```
is import path different based on the environment?

ColinPeppler replied:
yup I copied this from which supposedly makes it work with pytest

https://github.com/pytorch/pytorch/pull/130275/files

tenpercent replied:
it looks like `AOTIRunnerUtil` is defined in `test_aot_inductor_utils

ColinPeppler replied:
oh good catch!

tenpercent commented on torch/_inductor/codegen/cuda/cuda_kernel.py:
```diff
@@ -157,25 +166,55 @@ def call_kernel(
         wrapper = V.graph.wrapper_code
-        _, call_args, _, arg_types = self.args.python_argdefs()
+
+        if V.graph.cpp_wrapper:
```
Stylistic nit: looks like two `if` blocks can be merged?

ColinPeppler replied:
yes good call!

ColinPeppler commented on torch/_inductor/codegen/cuda/cuda_kernel.py:
```diff
@@ -157,25 +166,55 @@ def call_kernel(
+            # So, we replace with the real kernel name passed as an arg to this function.
+            self.signature = self.signature.replace("KERNEL_NAME", name)
+
+        # NOTE:
+        # We have to use python_argdefs even for cpp wrapper because python_argdefs
+        # gives us torch.dtype which is useful when calling
+        # generate_args_decl(call_args, arg_types) in generate_kernel_call()
```
this is an old comment, I must delete.

chenyang78 commented on torch/_inductor/codegen/cuda/cuda_kernel.py:
```diff
@@ -157,25 +166,55 @@ def call_kernel(
+        # gives us torch.dtype which is useful when calling
+        # generate_args_decl(call_args, arg_types) in generate_kernel_call()
+
+        if V.graph.cpp_wrapper:
```
Hmm, I feel like our convention is that we don't diverge with respect to cpp_wrapper in cuda_kernel.py. Instead, we rely on calling certain wrapper methods to differentiate cpp and non-cpp codegens. 

Adding @desertfire for more feedback.

desertfire replied:
@chenyang78 made a valid point. If the logic is complex, the convention is to separate them in the two wrapper codegen classes.

tenpercent commented on torch/_inductor/codegen/cpp_wrapper_cuda.py:
```diff
@@ -282,58 +282,78 @@ def generate_kernel_call(
+                    if config.abi_compatible:
+                        new_arg = f"var_{next(self.arg_var_id)}"
+                        self.writeline(
+                            f"auto* {new_arg} = get_data_ptr_wrapper({arg});"
```
After refactoring shapes out of template body the call_args may be integers

kadeng commented on torch/_inductor/async_compile.py:
```diff
@@ -213,10 +213,12 @@ def cpp_pybinding(self, argtypes: List[str], source_code: str):
 
         def task():
+            if aot_compile:
+                output_path, *_ = CUDACodeCache.compile(source_code, "o")
```
I would assume to ensure that a .o file is being built and inserted into the cache, even if dst_file_ext is, let's say "exe" or "so". So, if dst_file_ext is different from "o", it will mean that nvcc has to compile it all twice, which increases compilation time. Is this intentional?

ColinPeppler replied:
Yes, the doubled compilation time is intentional atm. One to compile a ".so" and another to compile a ".o". We need the ".so" for running Inductor and the ".o" for AOTInductor. I guess it wouldn't be too too bad as long as they're compiled on separate threads.

ColinPeppler committed:
Update on "[wip] enable CUTLASS in AOTI"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 amjames desertfire chauhang

[ghstack-poisoned]

ColinPeppler committed:
Update on "[wip] enable CUTLASS in AOTI"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 amjames desertfire chauhang

[ghstack-poisoned]

ColinPeppler commented on test/inductor/test_cutlass_backend.py:
```diff
@@ -184,8 +195,15 @@ def mm(a, b):
-            Y = mm(a, b)
+            Y = model(a, b)
+            if use_aoti:
+                Y_compiled = AOTIRunnerUtil.run(
```
We could pass in autotune_fallback_to_aten.

ColinPeppler committed:
Update on "[wip] enable CUTLASS in AOTI"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 amjames desertfire chauhang

[ghstack-poisoned]

ColinPeppler committed:
Update on "[wip] enable CUTLASS in AOTI"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 amjames desertfire chauhang

[ghstack-poisoned]

ColinPeppler committed:
Update on "[wip] enable CUTLASS in AOTI"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 amjames desertfire chauhang

[ghstack-poisoned]

ColinPeppler committed:
Update on "[wip] enable CUTLASS in AOTI"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 amjames desertfire chauhang

[ghstack-poisoned]

ColinPeppler committed:
Update on "[wip] enable CUTLASS in AOTI"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 amjames desertfire chauhang

[ghstack-poisoned]

ColinPeppler committed:
Update on "[wip] enable CUTLASS in AOTI"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 amjames desertfire chauhang

[ghstack-poisoned]

ColinPeppler committed:
Update on "[aotinductor] enable CUTLASS backend"

### Context
This PR allows CUTLASS kernels usage in AOTI. It does this by:
* For any CUTLASS kernels that win during autotuning, compile them as a .so & .o
* When creating the final model .so, link all the CUTLASS kernels .o files
* Make sure we codegen things correctly (argument dtypes and specify extern "C" linking for the CUTLASS kernel)

### Example
https://gist.github.com/ColinPeppler/e834fa2255c37e9444b6d540bf7bd04d#file-model-cpp-L548-L549

```
TORCH_LOGS="+output_code" python test/inductor/test_cutlass_backend.py -v -k test_max_autotune_cutlass_backend_regular_mm
```




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 amjames desertfire chauhang

[ghstack-poisoned]

ColinPeppler committed:
Update on "[aotinductor] enable CUTLASS backend"

### Context
This PR allows CUTLASS kernels usage in AOTI. It does this by:
* For any CUTLASS kernels that win during autotuning, compile them as a .so & .o
* When creating the final model .so, link all the CUTLASS kernels .o files
* Make sure we codegen things correctly (argument dtypes and specify extern "C" linking for the CUTLASS kernel)

### Example
https://gist.github.com/ColinPeppler/e834fa2255c37e9444b6d540bf7bd04d#file-model-cpp-L548-L549

```
TORCH_LOGS="+output_code" python test/inductor/test_cutlass_backend.py -v -k test_max_autotune_cutlass_backend_regular_mm
```




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 amjames desertfire chauhang

[ghstack-poisoned]

ColinPeppler commented:
I think this PR can be a bit cleaner, I plan to do a follow-up refactor.

tenpercent reviewed (approved):
Looks good, thanks for making it reusable for ROCm!

ColinPeppler commented:
@pytorchbot merge

ColinPeppler commented:
@pytorchbot merge

ColinPeppler commented:
@pytorchbot rebase

ColinPeppler committed:
Update

[ghstack-poisoned]

ColinPeppler commented:
@pytorchbot merge
